File: services/bandit/src/core/sync_conflicts.py
"""Sync-conflict detection (port of ``sync-conflicts.ts``)."""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any

from .slot import Intervals, overlaps_any


def find_conflicting_task_ids(
    fixed: Intervals, tasks: Sequence[dict[str, Any]]
) -> list[str]:
    if not fixed:
        return []
    return sorted(
        t["id"]
        for t in tasks
        if overlaps_any(
            fixed, t["startMs"], t["startMs"] + t["durationMinutes"] * 60_000
        )
    )

File: backend/src/scheduler/core/sync-conflicts.ts
import { overlapsAny, type Interval } from "./slot";

/**
 * Sync-conflict detection (issue #62 D) — pure. Given the fixed blocks a
 * timetable / exam / LMS sync just wrote and the user's own scheduled tasks,
 * returns the ids of the tasks now overlapping any of those blocks, sorted so
 * the same conflict set always yields the same list (stable dedupe key).
 */
export function findConflictingTaskIds(
  fixed: Interval[],
  tasks: { id: string; startMs: number; durationMinutes: number }[],
): string[] {
  if (fixed.length === 0) return [];
  return tasks
    .filter((t) =>
      overlapsAny(fixed, t.startMs, t.startMs + t.durationMinutes * 60_000),
    )
    .map((t) => t.id)
    .sort();
}

/** "X conflict(s)" wording used by the notification copy. */
export function conflictCountLabel(count: number): string {
  return `${count} conflict${count === 1 ? "" : "s"}`;
}
